pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_controller.go

```diff
@@ -46,124 +54,225 @@ func (r *HTTPRouteReconciler) Reconcile(ctx context.Context, req kube_ctrl.Reque
 			match[mesh_proto.ListenerTag] = string(*ref.SectionName)
 		}
 
-		selectors = append(selectors, &mesh_proto.Selector{
+		kumaRoute.Selectors = append(kumaRoute.Selectors, &mesh_proto.Selector{
 			Match: match,
 		})
 	}
 
-	var hostnames []string
+	routeConf, routeConditions, err := r.gapiToKumaRouteConf(ctx, mesh, route)
+	if err != nil {
+		return nil, nil, err
+	}
 
-	for _, hn := range route.Spec.Hostnames {
-		hostnames = append(hostnames, string(hn))
+	kumaRoute.Conf = routeConf
+
+	status := ParentConditions{}
+
+	for _, ref := range refRoutes {
+		status[ref] = routeConditions
 	}
 
-	var rules []*mesh_proto.GatewayRoute_HttpRoute_Rule
+	return kumaRoute, status, nil
+}
 
-	for _, rule := range route.Spec.Rules {
-		var backends []*mesh_proto.GatewayRoute_Backend
+func (r *HTTPRouteReconciler) shouldHandleParentRef(
+	ctx context.Context, route kube_client.Object, ref gatewayapi.ParentRef,
+) (bool, error) {
+	namespace := route.GetNamespace()
+	if ns := ref.Namespace; ns != nil {
+		namespace = string(*ns)
+	}
 
-		for _, backend := range rule.BackendRefs {
-			ref := backend.BackendObjectReference
+	gateway := &gatewayapi.Gateway{}
 
-			destination, err := r.gapiToKumaRef(ctx, mesh, namespace, ref)
-			if err != nil {
-				return nil, err
-			}
+	if *ref.Group != gatewayapi.Group(gatewayapi.GroupName) || *ref.Kind != gatewayapi.Kind(reflect.TypeOf(gateway).Elem().Name()) {
+		return false, nil
+	}
 
-			backends = append(backends, &mesh_proto.GatewayRoute_Backend{
-				// Weight has a default of 1
-				Weight:      uint32(*backend.Weight),
-				Destination: destination,
-			})
+	if err := r.Client.Get(ctx, kube_types.NamespacedName{Namespace: namespace, Name: string(ref.Name)}, gateway); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			return false, nil
+		} else {
+			return false, err
 		}
+	}
+
+	class, err := getGatewayClass(ctx, r.Client, gateway.Spec.GatewayClassName)
+	if err != nil {
+		return false, err
+	}
+
+	return class.Spec.ControllerName == controllerName, nil
+}
 
-		var matches []*mesh_proto.GatewayRoute_HttpRoute_Match
+// reconcileLabelledObjectSet manages a set of owned objects based on labeling
+// with the owner key and indexed by the service tag and listener tag.
+func reconcileLabelledObject(
+	ctx context.Context,
+	client kube_client.Client,
+	owner kube_types.NamespacedName,
+	routeSpec *mesh_proto.GatewayRoute,
+) error {
+	// First we list which existing objects are owned by this route.
+	// We expect either 0 or 1 and depending on whether routeSpec is nil
+	// we either create an object or update or delete the existing
+	ownerLabelValue := fmt.Sprintf("%s-%s", owner.Namespace, owner.Name)
+	labels := kube_client.MatchingLabels{
+		ownerLabel: ownerLabelValue,
+	}
+
+	list := &mesh_k8s.GatewayRouteList{}
+	if err := client.List(ctx, list, labels); err != nil {
+		return err
+	}
+
+	if len(list.Items) > 1 {
+		return fmt.Errorf("expected either zero or one owned items")
```

Comment:
If this happens, the operator is not likely to know how to response. Maybe file an issue to improve this?

Reply:
I've added more info that it's an internal error.